### PR TITLE
incremental checking and imports

### PR DESCRIPTION
**Pull request by ivoysey**
this PR is a *draft* because it is **not** ready to be merged or reviewed. this will ultimately address #110 and #81 when the dust settles.

i've made a PR so that i can both keep my branch up to date with the changes in master to avoid a big bang integration with lots of merge conflicts later on, but also make it easy to see at a glance which changes are unique to this branch and which are from other branches by way of master.

for now, the code in my branch may exist in a messy and non-compiling state because i'm in the process of adapting some code from `redtt` that relies on a similar but not quite the same environment. once i get past an initial prefix of that, i will be able to maintain the good hygiene invariant of keeping my branch compiling, but i'm just not there yet. i'm aware of this situation and i think it's for good reasons, so do not be alarmed.

i will request reviews and ask questions of specific people as i have them, either here or on the related issues!

### CODE REVIEW

**Comment by favonia** [comment]
It seems you are getting started! I would like to see more code. I suggest you pay attention to the role of `Contextual` in `RotIO` in redtt and find an equivalent way to do it using `ElabEnv` in cooltt. In particular, pay attention to how `json_of_name` was implemented.

**Comment by ivoysey** [comment]
@favonia OK! I'll take that as a good sign and keep hacking today, see where I get, and bug you again later. Thanks!

File: src/lib/SlushIO.ml [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+      tm_of_json tm3 >>= fun tm3 ->
+      tm_of_json tm4 >>= fun tm4 ->
+      tm_of_json tm5 >>= fun tm5 ->
+      ret @@ VProj(tm1, tm2, tm3, tm4, tm5)
+
+    | `A [`String "CodeExt"; i; tm1; glo; tm2] ->
+      tm_of_json tm1 >>= fun tm1 ->
+      global_of_json glo >>= fun glo ->
+      tm_of_json tm2 >>= fun tm2 ->
+      ret @@ CodeExt(int_of_json i, tm1, glo, tm2) (* todo *)
```
Comment (jonsterling) [inline src/lib/SlushIO.ml:476]: Just a small stylistic comment, in ML languages we think of the argument to such a constructor as a tuple --- so we prefer to write `CodeExt (...)` with a space.

Comment (ivoysey) [reply src/lib/SlushIO.ml:476]: do you have feelings about single argument constructors? e.g. `ElIn (tm)` (uniform but redundant) vs `ElIn(tm)` (unlikely) or `ElIn tm` (concise but not uniform)?

Comment (jonsterling) [reply src/lib/SlushIO.ml:476]: Since most languages don’t have a type of unary tuples , so I think we are safe with the latter haha

Comment (ivoysey) [reply src/lib/SlushIO.ml:476]: got it. i shall endeavor to be more jon-friendly in this respect!

File: src/basis/Monad.ml [hunk 2]
```diff
@@ -19,6 +19,9 @@ module type Notation = sig
   val (|>>) : 'a m -> ('a -> 'b m) -> 'b m
   val (@<<) : ('a -> 'b m) -> 'a m -> 'b m
   val (<&>) : 'a m -> 'b m -> ('a * 'b) m
+  val (>>=) : 'a m -> ('a -> 'b m) -> 'b m
+  val (<@>>) : ('a -> 'b) -> 'a m -> 'b m
```
Comment (jonsterling) [inline src/basis/Monad.ml:23]: Btw I don't know if it helps for me to mention this, but these two operators are already present a few lines above --- with different names.

Comment (ivoysey) [reply src/basis/Monad.ml:23]: oh, good point. i was mostly just matching ASCII art to get things to type check last week, but that's a redundancy that needn't exist. i'll clean it up.

**Pull request closed without merging**